### DISCUSSION TITLE
question: why lwm2m_rd_client_stop might block

### DISCUSSION THREAD

**Discussion by chmielewskiandreas** [zephyrproject-rtos/zephyr#54478] [Q&A]
routine lwm2m_rd_client_stop of lwm2m_rd_client.c might block until engine is IDLE. This is in particular bad if deregister is set to true and there are some issues with deregistration e.g. timeout during reregistration. The worst thing that can happen is that you block your work queue, where you call that function from, totally. This is the change:

```
commit 43c988d43e70e7234940b1910a27b5cf0b87bb9b
Author: Seppo Takalo <seppo.takalo@nordicsemi.no>
Date:   Wed Mar 30 15:24:37 2022 +0300

    net: lwm2m: Clean up context on stop

    When lwm2m_rd_client_stop() was called and immediately
    followed by lwm2m_rd_client_start() it leaked the file
    handle for existing socket.

    Problem can be fixed when rd_client_stop() does not
    move state machine to IDLE, but instead DEREGISTER
    and then allow state machine to move forward.

    I added a blocking wait for rd_client_stop() because
    it needs to wait for proper clean up.

    I also move couple of lwm2m_engine_context_close() to
    set_sm_state() event handler or similarly in lwm2m_engine.c
    there was couple of places where context was not properly
    cleaned.

    Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>
```

This blocking call causes some troubles for us as as you can imagine the caller might block for ever/some time. I can not follow the reason why this code was introduced and how to fix this in a way that it solves the origin issue. Can't we introduce a new RD_CLIENT_EVENT_DEREGISTERED for this? There is already RD_CLIENT_EVENT_DEREGISTER_FAILURE so I actually assume that such an event would make sense.

**Comment by SeppoTakalo** [comment]
I think there are no real reason any more.
The state machine has been a quite fragile in the past, but it has got much more testing now.

Maybe the loop was introduced while developing, just as an extra safety measure, and once it was working, I did not remove it.

I'm proposing to remove it in https://github.com/zephyrproject-rtos/zephyr/pull/56564 as I don't currently see any reason for it. The loop itself does not do any corrective action, so if there was any reason that might cause state machine to not go into IDLE, it would block forever. Many corner cases has been fixed in the state machine since introduction of the loop, so I believe it does not serve any purpose anymore.
